src/safeposix/dispatcher.rs

```diff
@@ -152,6 +167,159 @@ pub extern "C" fn dispatcher(cageid: u64, callnum: i32, arg1: Arg, arg2: Arg, ar
+                let rv = check_and_dispatch!(cage.recvfrom_syscall, interface::get_int(arg1), interface::get_mutcbuf(arg2), interface::get_usize(arg3), interface::get_int(arg4), Ok::<&mut Option<&mut interface::GenSockaddr>, i32>(&mut Some(&mut addr)));
+
+                if rv >= 0 {
+                    interface::copy_out_sockaddr(arg5, arg6, addr);
+                }
+                rv
+            } else {
+                syscall_error(Errno::EINVAL, "recvfrom", "exactly one of the last two arguments was zero")
+            }
+        }
+
+        CONNECT_SYSCALL => {
+            let addrlen = get_onearg!(interface::get_uint(arg3));
+            let addr = get_onearg!(interface::get_sockaddr(arg2, addrlen));
+            check_and_dispatch!(cage.connect_syscall, interface::get_int(arg1), Ok::<&interface::GenSockaddr, i32>(&addr))
+        }
+        LISTEN_SYSCALL => {
+            check_and_dispatch!(cage.listen_syscall, interface::get_int(arg1), interface::get_int(arg2))
+        }
+        ACCEPT_SYSCALL => {
+            let mut addr = interface::GenSockaddr::V4(interface::SockaddrV4::default()); //value doesn't matter
+            let nullity1 = interface::arg_nullity(&arg2);
+            let nullity2 = interface::arg_nullity(&arg3);
+            
+            if nullity1 && nullity2 {
+                check_and_dispatch!(cage.accept_syscall, interface::get_int(arg1), Ok::<&mut interface::GenSockaddr, i32>(&mut addr))
+            } else if !(nullity1 || nullity2) {
+                let rv = check_and_dispatch!(cage.accept_syscall, interface::get_int(arg1), Ok::<&mut interface::GenSockaddr, i32>(&mut addr));
+                if rv >= 0 {
+                    interface::copy_out_sockaddr(arg2, arg3, addr);
+                }
+                rv
+            } else {
+                syscall_error(Errno::EINVAL, "accept", "exactly one of the last two arguments was zero")
+            }
+        }
+
+        GETPEERNAME_SYSCALL => {
+            let mut addr = interface::GenSockaddr::V4(interface::SockaddrV4::default()); //value doesn't matter
+            if interface::arg_nullity(&arg2) || interface::arg_nullity(&arg3) {
+                return syscall_error(Errno::EINVAL, "getpeername", "Either the address or the length were null");
+            }
+            let rv = check_and_dispatch!(cage.getpeername_syscall, interface::get_int(arg1), Ok::<&mut interface::GenSockaddr, i32>(&mut addr));
+
+            if rv >= 0 {
+                interface::copy_out_sockaddr(arg2, arg3, addr);
+            }
+            rv
+        }
+        GETSOCKNAME_SYSCALL => {
+            let mut addr = interface::GenSockaddr::V4(interface::SockaddrV4::default()); //value doesn't matter
+            if interface::arg_nullity(&arg2) || interface::arg_nullity(&arg3) {
+                return syscall_error(Errno::EINVAL, "getsockname", "Either the address or the length were null");
```

Comment:
Aren't all addresses inherently being tested for being null in native_client during NaClUserToSysAddrRange? Why do we have to check for address validity again at Safeposix?

Reply:
They are being tested within NaCl but it is good sanity checking to make sure that any pointer that we have is checked for nullity before we use it, plus it's generally good practice.